### PR TITLE
Add additional test coverage for the regexp contained in RepositoryNameRegexp

### PR DESCRIPTION
**Pull request by aaronlehmann**
This was inspired by problems found with new regexps proposed in PR #690

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

### CODE REVIEW

**Comment by aaronlehmann** [comment]
@stevvooe PTAL

File: registry/api/v2/names_test.go [hunk 1]
```diff
@@ -149,3 +149,143 @@ func TestRepositoryNameRegexp(t *testing.T) {
...
+		{
+			input: "a/aa/a",
+		},
+		{
+			input:   "foo.com/",
+			invalid: true,
+		},
+		{
+			// currently not allowed by the regex
+			input:   "foo.com:8080/bar",
```
Comment (dmcgowan) [inline registry/api/v2/names_test.go]: This seems problematic

File: registry/api/v2/names_test.go [hunk 2]
```diff
@@ -149,3 +149,143 @@ func TestRepositoryNameRegexp(t *testing.T) {
...
 	}
 }
+
+func TestRepositoryNameRegexp(t *testing.T) {
+	for _, testcase := range []struct {
+		input   string
+		invalid bool
+	}{
+		{
+			input: "short",
```
Comment (stevvooe) [inline registry/api/v2/names_test.go]: No need to repeat this list from the other test. Just use the presence of an error to infer validity.

File: registry/api/v2/names_test.go (hunk 1 above)
Comment (stevvooe) [inline registry/api/v2/names_test.go]: @aaronlehmann Just add a TODO here indicating this should be valid when we move the `reference` package.

File: registry/api/v2/names_test.go [hunk 3]
```diff
@@ -6,7 +6,7 @@ import (
 	"testing"
 )
 
-func TestRepositoryNameRegexp(t *testing.T) {
+func TestRepositoryComponentNameRegexp(t *testing.T) {
```
Comment (stevvooe) [inline registry/api/v2/names_test.go]: This is actually testing `ValidateRepositoryName`.

**Comment by aaronlehmann** [comment]
Unified the testcases.

File: registry/api/v2/names_test.go [hunk 4]
```diff
@@ -6,14 +6,23 @@ import (
...
+	// regexpTestcases is a unified set of testcases for
+	// TestValidateRepositoryName and TestRepositoryNameRegexp.
+	// Some of them are valid inputs for one and not the other.
+	regexpTestcases = []struct {
+		// input is the repository name or name component testcase
 		input string
-		err   error
+		// errFromValidateRepositoryName is the error expected from
+		// ValidateRepositoryName, or nil
+		errFromValidateRepositoryName error
```
Comment (stevvooe) [inline registry/api/v2/names_test.go]: Please see [Rule 6 in coding style](https://github.com/docker/distribution/blob/master/CONTRIBUTING.md#coding-style).

**Comment by aaronlehmann** [comment]
Renamed variables, and merging with preemptive approval from @stevvooe.

**Pull request merged**